x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java

```diff
@@ -63,24 +64,26 @@ public Weight doCache(Weight weight, QueryCachingPolicy policy) {
      */
     static boolean cachingIsSafe(Weight weight, IndicesAccessControl.IndexAccessControl permissions) {
         // support caching for common queries, by inspecting the field
-        // TODO: If in the future there is a Query#extractFields() then we can do a better job
-        Set<String> fields = new HashSet<>();
         try {
-            FieldExtractor.extractFields(weight.getQuery(), fields);
-        } catch (UnsupportedOperationException ok) {
-            // we don't know how to safely extract the fields of this query, don't cache.
-            return false;
-        }
+            weight.getQuery().visit(new QueryVisitor() {
+                @Override
+                public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
+                    return this; // we want to use the same visitor for must_not clauses too
+                }
 
-        // we successfully extracted the set of fields: check each one
-        for (String field : fields) {
-            // don't cache any internal fields (e.g. _field_names), these are complicated.
-            if (field.startsWith("_") || permissions.getFieldPermissions().grantsAccessTo(field) == false) {
-                return false;
-            }
+                @Override
+                public boolean acceptField(String field) {
+                    // don't cache any internal fields (e.g. _field_names), these are complicated.
+                    if (field.startsWith("_") || permissions.getFieldPermissions().grantsAccessTo(field) == false) {
+                        throw new UnsupportedOperationException();
+                    }
+                    return super.acceptField(field);
+                }
+            });
+        } catch (UnsupportedOperationException e) {
+            return false;
```

Comment:
I've checked that the new visitor strategy is equivalent for all the query types we previously supported caching while FLS is set.

However, the previous strategy effectively had an allowlist of query types for caching with FLS, whereas this new strategy is more lax. I would lean on maintaining an allow list of query types. What do you think?
I'm not sure that all the queries that call `QueryVisitor#visitLeaf` and don't call `QueryVisitor#acceptField` are safe to cache in an FLS context.